packages/ember-data/lib/system/model/states.js

```diff
@@ -343,7 +345,7 @@ var DirtyState = {
 
     rolledBack: function(record) {
       get(record, 'errors').clear();
-      record.triggerLater('ready');
+      record.transitionTo('loaded.saved');
```

Comment:
This will break the case when a new record is saved, has errors and is then rolled back. By transitioning into the `loaded.saved` state, this record is now marked as created, but it should be deleted instead...

Failing test case:

``` javascript
test("invalid new record can be rollbacked", function() {
  var person;
  var adapter = DS.RESTAdapter.extend({
    ajax: function(url, type, hash) {
      var adapter = this;

      return new Ember.RSVP.Promise(function(resolve, reject) {
        /* If InvalidError is passed back in the reject it will throw the
           exception which will bubble up the call stack (crashing the test)
           instead of hitting the failure route of the promise.
           So wrapping the reject in an Ember.run.next makes it so save
           completes without failure and the failure hits the failure route
           of the promise instead of crashing the save. */
        Ember.run.next(function() {
          reject(adapter.ajaxError({ name: 'is invalid' }));
        });
      });
    },

    ajaxError: function(jqXHR) {
      return new DS.InvalidError(jqXHR);
    }
  });

  env = setupStore({ person: Person, adapter: adapter });

  run(function() {
    person = env.store.createRecord('person', { id: 1 });
  });

  equal(person.get('isNew'), true, "must be new");
  equal(person.get('isDirty'), true, "must be dirty");

  run(function() {
    person.save().then(null, async(function() {
      equal(person.get('isValid'), false);
      person.rollback();

      // assertions taken from the test 'new record can be rollbacked' http://git.io/vTsKR
      equal(person.get('isNew'), false, "must not be new");
      equal(person.get('isDirty'), false, "must not be dirty");
      equal(person.get('isDeleted'), true, "must be deleted");
    }));
  });
});
```